### PR TITLE
## Release 2.0.0

### PR DESCRIPTION
**Pull request by golluroSICKAG**
# Version 2.0.0

## New features
- Supports FlowConfig feature to receive data / forward results
- Added new syntaxes: 'stringPack', 'stringUnpack', 'searchAndCut', 'dateTime', 'timestamp'
- New feature to merge multiple parameters of different events and to provide them within a single event with multiple parameters (see 'setMergeData' and 'OnNewData_EXPRESSION')
- New feature to provide the result of the expression criteria ('OnNewCriteriaResult_EXPRESSION')
- Provide version of module via 'OnNewStatusModuleVersion'
- Function 'getParameters' to provide PersistentData parameters
- Check if features of module can be used on device and provide this via 'OnNewStatusModuleIsActive' event / 'getStatusModuleActive' function
- Function to 'resetModule' to default setup

## Improvements
- extended 'addExpression' with attribute to set if it should only merge data
- proper UI table selection
- Improved behaviour for loading parameters (not needed to wait for other modules)
- New UI design available (e.g. selectable via CSK_Module_PersistentData v4.1.0 or higher), see 'OnNewStatusCSKStyle'
- 'loadParameters' returns its success
- 'sendParameters' can control if sent data should be saved directly by CSK_Module_PersistentData
- Added UI icon and browser tab information

## Bugfix
- Check if CROWN really provides XML content before trying to use it
- UI table selection runs into error for some selections
- Return value of 'OnNewResult_...' event was always 'bool' but needs to be 'auto'

## Following tasks were checked

- [x] Module is coded and structured according to the "Developing guideline for modules" described within the CSK documentation
- [x] All functions/events/parameters are documented within the manifest documentation
- [x] The manifest description of the main CROWN includes main information about the purpose of the module and how to use it in general
- [x] API docu based on the manifest was created and stored within the "docu"-folder of the repository
- [x] Internal LUA code documentation exists for variables and non served functions
- [x] All relevant infos are logged via the SharedLogger 'ModuleLogger'
- [x] Module supports persistent data feature based on 'CSK_Module_PersistentData'
- [x] Module supports user management based on 'CSK_Module_UserManagement'
- [x] No open "ToDos" within the code or at least clearly explained comments why they exist...
- [x] "Version" key in app manifest was updated following semantic versioning (and use '0.x.y' for test / experimental modules which are not yet ready to be officially released)
- [x] Meaningful IDs used for UI elements
- [x] Module was tested on an AppSpace device (at least on SICK AppEngine) with no error message
- [x] README.md is up to date (incl. info of device + firmware the module was tested with)
- [x] CHANGELOG.md is up to date

### CODE REVIEW

File: CSK_Module_ResultManager/scripts/Data/ResultManager/FlowConfig/ResultManager_OnNewResult.lua [hunk 1]
```diff
@@ -0,0 +1,77 @@
+-- Block namespace
```
Comment (kogliinSICKAG) [inline CSK_Module_ResultManager/scripts/Data/ResultManager/FlowConfig/ResultManager_OnNewResult.lua:1]: I tried to run the demo flows and tested it with the different resources of the ImagePlayer. When the DeepLearning images are used, the ResultManager runs into this error:
[09:12:09.216: WARNING: API: App 'CSK_Module_ResultManager': Pos 'Data/ResultManager/helper/Json.lua:130': API 'anonym_3_CSK_Module_ResultManager:Data/ResultManager/ResultManager_Controller.lua:191'] Runtime error: unexpected type 'userdata'

File: CSK_Module_ResultManager/pages/pages/CSK_Module_ResultManager/CSK_Module_ResultManager.html [hunk 2]
```diff
@@ -1,425 +1,925 @@
...
-			<crown-binding property="value" name="CSK_ResultManager/pageCalled"
-				path="result/data/empty" auto-update="0" update-on-resume>
-			</crown-binding>
-		</stacked-view>
-	</layout-column>
-	<layout-column id="ColumnLayout2" style="align-items: stretch">
-	</layout-column>
-  
-</layout-row>
+<layout-row id="RowLayout1" class="myCustomBackground_CSK_Module_ResultManager">
```
Comment (kogliinSICKAG) [inline CSK_Module_ResultManager/pages/pages/CSK_Module_ResultManager/CSK_Module_ResultManager.html:1]: The ResultManager GUI does not enable the settings for PersistentData except for the checkbox of the flow config priority. I also tried to uncheck the flow config priority but the settings still remained disabled.
image
Maybe this is due to the runtime error within json.lua?

**Review by kogliinSICKAG — COMMENTED** [review]
As discussed, maybe the error with userdata withi json.lua has been due to previous testing of the flow config.

File: CSK_Module_ResultManager/scripts/Data/ResultManager/FlowConfig/ResultManager_OnNewResult.lua (hunk 1 above)
Comment (golluroSICKAG) [reply CSK_Module_ResultManager/scripts/Data/ResultManager/FlowConfig/ResultManager_OnNewResult.lua:1]: Please try to reproduce the issue with the updated modules. Thx!

File: CSK_Module_ResultManager/pages/pages/CSK_Module_ResultManager/CSK_Module_ResultManager.html (hunk 2 above)
Comment (golluroSICKAG) [reply CSK_Module_ResultManager/pages/pages/CSK_Module_ResultManager/CSK_Module_ResultManager.html:1]: I was not able to reproduce this issue.
Is it possible, that the PersistentData module was not running? If so, than it is intended to look like this.

**Comment by golluroSICKAG** [comment]
Thx for the review so far!

**Pull request merged**